File: rust/src/y2015/d04.rs
use md5::Digest;

use crate::util::format_soln_string;

fn get_key(n: u32, base: &str) -> String {
    format!("{}{}", base, n.to_string())
}

fn digest_leading_zeroes(d: Digest, n: usize) -> bool {
    // todo: could be refactored to not use format and instead compare bytes
    format!("{:x}", d).starts_with("0".repeat(n).as_str())
}

fn find_soln(leading_zeroes: usize, input: &String) -> u32 {
    let mut n = 1;
    let mut key = get_key(n, &input);
    let mut digest = md5::compute(key.as_bytes());
    while !digest_leading_zeroes(digest, leading_zeroes) {
        n += 1;
        key = get_key(n, &input);
        digest = md5::compute(key.as_bytes());
    }
    return n;
}

pub fn soln(input: String) -> String {
    let soln1 = find_soln(5, &input);
    let soln2 = find_soln(6, &input);
    return format_soln_string(
        String::from(soln1.to_string()),
        String::from(soln2.to_string()),
    );
}

File: js/y2015/d04.js
const { fmtAnsWithRuntime } = require('../util.js');
const { md5 } = require("../util/md5.js");

/**
 * @param {string} firstChars 
 */
function findN(prefix, firstChars) {
  let n = 0;
  let hash = "";
  while (!hash.startsWith(firstChars)) {
    n++;
    hash = md5(`${prefix}${n}`);
  }
  return n;
}

/**
 * @param {string} rawInput
 */
function partOne(input) {
  return findN(input, "00000");
}

/**
 * @param {string} rawInput
 */
function partTwo(input) {
  return findN(input, "000000");
}

function soln(rawInput) {
  fmtAnsWithRuntime(
    () => partOne(rawInput),
    () => partTwo(rawInput)
  );
}

module.exports = { soln };
